### PR TITLE
fixed send invitation email

### PR DESCRIPTION
**Pull request by Panioglo**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -8,9 +8,16 @@ def new
...
-    after_login_redirect
+    user = User.find_by_email(user_params[:email])
+    unless user.nil?
+      if user.confirmation_token.nil?
+        user = authenticate!(:password)
+        set_website(user.default_website)
+        after_login_redirect
+      else
+        redirect_to login_url, alert: "Confirm your email first"
+      end
```
Comment (Chocksy) [inline app/controllers/sessions_controller.rb]: all this code should be in warden. This way you can do a very simple check in there to see if it's confirmed or not and send a different message.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,9 +1,19 @@
 class UsersController < ApplicationController
-  layout "landing"
+  layout 'landing', :only => [:new, :create]
   skip_before_action :authenticate!
 
   def index; end
 
+  def edit
+    @user = current_user
+  end
```
Comment (Chocksy) [inline app/controllers/users_controller.rb]: use `load_and_authorize only: [:edit, :update]` so you won't need the `@user = current_user` defined in the edit.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -14,6 +14,12 @@ class ApplicationController < ActionController::Base
     redirect_to login_url, :alert => exception.message
   end
 
+  def reload_page
+    respond_to do |format|
+      format.js { render inline: 'location.reload();' }
+    end
+  end
```
Comment (Chocksy) [inline app/controllers/application_controller.rb]: i want you to remove this method there is no place for this stupid thing. We should only use `redirect_to` with flash messages and not get back to js reloading it's not good.

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,9 +1,19 @@
...
+
+  def update
+    @user = current_user
+    @user.update_attributes(user_params)
+    reload_page
```
Comment (Chocksy) [inline app/controllers/users_controller.rb]: Here you need to add a check and show errors if the update of the form fails

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -14,15 +24,35 @@ def new
...
     user = User.new(user_params)
-    
+
     authenticate!(:password) if user.save
+    unless params[:invitation_token].blank?
+      @website_member = WebsiteMember.find_by_invitation_token(params[:invitation_token])
+      unless @website_member.nil?
+        @website_member.update_attributes(user_id: user.id, role: 2)
+        set_website(Website.find(@website_member.website_id))
+      end
```
Comment (Chocksy) [inline app/controllers/users_controller.rb]: this code should be in the user model and call this by doing `user.accept(params[:invitation_token])`

File: app/controllers/website_members_controller.rb [hunk 6]
```diff
@@ -1,14 +1,12 @@
...
+    @website_members = current_website.website_members.where("user_id IS NOT NULL")
   end
 
   def destroy
     @website_member.destroy
     return false unless @website_member.errors.full_messages.blank?
-    respond_to do |format|
-      format.js {render inline: 'location.reload();' }
-    end
+    reload_page
```
Comment (Chocksy) [inline app/controllers/website_members_controller.rb]: use redirect_to here instead of the reload_page method. We should always know where to redirect after getting into an action.

File: app/controllers/websites_controller.rb [hunk 7]
```diff
@@ -7,9 +7,7 @@ def index
 
   def update
     @website.update_attributes(website_params)
-    respond_to do |format|
-      format.js { render inline: 'location.reload();' }
-    end
+    reload_page
```
Comment (Chocksy) [inline app/controllers/websites_controller.rb]: use redirect_to

File: app/models/user.rb [hunk 8]
```diff
@@ -7,6 +9,9 @@ class User < ActiveRecord::Base
   validates :password, confirmation: true, on: :create
   has_secure_password(validations: false)
 
+  before_create { generate_token(:uid) if Proc.new { |u| u.provider == 'email'} }
```
Comment (Chocksy) [inline app/models/user.rb]: why are you doing a proc here? The one below doesn't have a proc and it works...

File: app/views/users/confirm_account.json.jbuilder [hunk 9]
```diff
@@ -0,0 +1 @@
+json.user(@user, :id, :confirmation_token, :confirmation_sent_at, :confirmed_at)
```
Comment (Chocksy) [inline app/views/users/confirm_account.json.jbuilder]: what is this file for?

File: app/views/users/edit.html.haml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                %td
+                  %h5 #{@user.email}
+              %tr
+                %td
+                  %p Provider:
+                %td
+                  %h5 #{@user.provider}
+              %tr
+                %td
+                  %p Created at: &nbsp&nbsp&nbsp
```
Comment (Chocksy) [inline app/views/users/edit.html.haml]: stop using `&nbsp`

File: app/views/users/new.html.haml [hunk 11]
```diff
@@ -27,6 +27,7 @@
               .input-group-addon
                 %span.glyphicon.glyphicon-lock
               = f.password_field :password_confirmation, :placeholder => 'Confirm password', class: 'form-control input-lg', required: true
+              = hidden_field_tag(:invitation_token, params[:id])
```
Comment (Chocksy) [inline app/views/users/new.html.haml]: put the hidden filed at the top of the form

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -34,11 +34,14 @@ def current_website
 
   def set_website(website)
     return unless logged_in?
-    @website = website
-    session[:epiclogger_website_id] = @website.id
+    unless website.nil?
```
Comment (Chocksy) [inline app/controllers/application_controller.rb]: use `return unless website.nil?` one line above the 2 that were below. So we can keep a clean look to the code.

File: app/controllers/invites_controller.rb [hunk 13]
```diff
@@ -0,0 +1,54 @@
+class InvitesController < ApplicationController
+  skip_before_action :authenticate!, only: :accept
+
+  def create
+    begin
+      unless Invite.where('email = ? AND website_id = ? AND invited_by_id =?', invite_params[:email], current_website.id, current_user.id).blank?
```
Comment (Chocksy) [inline app/controllers/invites_controller.rb]: do it like this:

```
raise 'User already invited' unless current_website.invites.where(email: invite_params[:email]).blank?
```

As we don't need to do the full check including the invited_by_id that is redundant.

File: app/controllers/invites_controller.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        raise "Duplicate invites"
+      end
+      invite = current_website.invites.new(invited_by_id: current_user.id, email: invite_params[:email])
+      UserMailer.member_invitation(invite.id).deliver_later if invite.save
+      redirect_to new_invite_url, notice: 'Email sent'
+
+    rescue Exception => e
+      redirect_to(:action => 'new', notice: e.message)
+    end
+  end
```
Comment (Chocksy) [inline app/controllers/invites_controller.rb:9]: Use included exception rescue here into the method that means you should do this:

```
def
...
rescue => e
...
end
```

Without begin and without double ends.

There is also another way of doing this without rescue and that is by doing the redirect above with one check. Like so:

```
redirect_to(:new_invite, notice: 'User already invited') && return unless current_website.invites.where(email: invite_params[:email]).blank?
```

If you add the and return at the end then it's going to stop execution.

File: app/controllers/invites_controller.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+  def new; end
+
+  def show
```
Comment (Chocksy) [inline app/controllers/invites_controller.rb]: this method seems to be similar or close to what i made below with the accept method so i think you can remove it. Tell me if it's not the case so we can discuss.

File: app/controllers/website_members_controller.rb [hunk 16]
```diff
@@ -1,14 +1,12 @@
 class WebsiteMembersController < ApplicationController
 	load_and_authorize_resource
   def index
-    @members = current_website.website_members
+    @website_members = current_website.website_members.where("user_id IS NOT NULL")
```
Comment (Chocksy) [inline app/controllers/website_members_controller.rb]: why do we need this where here? Are we creating website_members without user_id??? ... That should not be the case.

File: app/mailers/user_mailer.rb [hunk 17]
```diff
@@ -1,6 +1,11 @@
 class UserMailer < ApplicationMailer
   default from: "Epic Logger <admin@epiclogger.com>"
 
+  def email_confirmation(token)
```
Comment (Chocksy) [inline app/mailers/user_mailer.rb]: this seems wrong and it should not be done this way.
The way to call emails is by doing this with action_mailer: `UserMailer.email_confirmation(user).deliver_later` so notice there is the full user record sent. That's how this should work as well. So the only thing you would do in this method would be a `@user = user` to assign the variable for mailer view.

File: app/mailers/user_mailer.rb [hunk 18]
```diff
@@ -9,12 +14,11 @@ def notify_subscriber(group, user, message)
     mail(to: @user.email, subject: 'Epic Logger Subscriber notification')
   end
 
-  def member_invitation(website_id, email, website_member_id, inviter_id)
-    @user = User.find(inviter_id)
-    @website = Website.find(website_id)
-    @token = WebsiteMember.find(website_member_id).invitation_token
-    @email = email
-    mail(to: @email, subject: 'Epic Logger Invite Users')
+  def member_invitation(invite_id)
```
Comment (Chocksy) [inline app/mailers/user_mailer.rb]: considering what i wrote above please send the full invite record here and not the id. If you are using deliver_later and action mailer.

File: app/controllers/website_members_controller.rb (hunk 16 above)
Comment (Panioglo) [inline app/controllers/website_members_controller.rb]: We can remove this. Now we have invites table and there is no case when website_member is created without an user_id.

File: app/models/user.rb [hunk 19]
```diff
@@ -28,4 +33,10 @@ def self.create_with_omniauth(auth)
       user.email = auth["info"]["email"]
     end
   end
+
+  def send_confirmation
+    generate_token(:confirmation_token)
+    self.confirmation_sent_at = Time.now
+    UserMailer.email_confirmation(self.confirmation_token).deliver_later
```
Comment (Chocksy) [inline app/models/user.rb]: so considering what i said above about the user_email here you should send `self` instead of the token.

File: app/mailers/user_mailer.rb (hunk 17 above)
Comment (Chocksy) [inline app/mailers/user_mailer.rb]: Also here add a `return if user.confirmed?` by making a `def confirmed?` method in the user model to check if the user is confirmed. That will stop the email sending if the user was confirmed.

File: app/controllers/invites_controller.rb (hunk 14 above)
Comment (Panioglo) [inline app/controllers/invites_controller.rb:9]: I didn't prefered to use return since last 2 times when i wanted to return false but the execusion still continued.. maybe it was just a bad experience

File: app/views/invites/new.html.haml [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+  .col-md-12.pagecontainer
+    .pagetitle
+      %h3 Add member to your website
+    %br/
+    .row
+      .col-md-6
+        .panel.panel-default
+          .panel-heading Send an email invitation to:
+          .panel-body
+            = simple_form_for Invite.new, url: invites_path, method: :create, wrapper: :vertical_form do |f|
```
Comment (Chocksy) [inline app/views/invites/new.html.haml]: you don't need url and method here. It should work as expected.

File: app/views/users/edit.html.haml [hunk 21]
```diff
@@ -0,0 +1,45 @@
+%br
+.row
+  .col-md-12.pagecontainer
+    = simple_form_for @user, url: user_path(id: @user.id), method: :put, remote: true, wrapper: :vertical_form do |f|
```
Comment (Chocksy) [inline app/views/users/edit.html.haml]: same here no need for url and method.

File: config/initializers/warden.rb [hunk 22]
```diff
@@ -25,34 +25,34 @@ def valid?
 
   def authenticate!
     user = User.find_by_email(user_params[:email])
-    if user && user.authenticate(user_params[:password])
+    if user && user.authenticate(user_params[:password]) && user.confirmation_token.nil?
```
Comment (Chocksy) [inline config/initializers/warden.rb]: here if you make that method `confirmed?` that i told you about you can write this as `if user && user.authenticate(user_params[:password]) && user.confirmed?` and it's going to be good.

File: config/routes.rb [hunk 23]
```diff
@@ -25,8 +25,12 @@
       get :wizard_install
     end
   end
-  resources :users, only: [:index, :edit]
-  resources :invitations, only: [:new, :show]
+  resources :users, only: [:index, :edit, :update] do
+    member do
+      get :confirm_account
```
Comment (Chocksy) [inline config/routes.rb]: you can write this as `get :confirm` as there is no need for `confirm_account`. It's understandable.

File: app/controllers/invites_controller.rb (hunk 14 above)
Comment (Chocksy) [inline app/controllers/invites_controller.rb:9]: That's because you had to use the `()` for redirect_to or if you don't use them you need to replace `&&` with `and` for it to work.

File: app/controllers/website_members_controller.rb (hunk 16 above)
Comment (Chocksy) [inline app/controllers/website_members_controller.rb]: ok. Remove it then.

File: app/controllers/invites_controller.rb [hunk 24]
```diff
@@ -39,10 +19,10 @@ def accept
       @invite.accept(current_user)
       set_website(@invite.website)
       # it will redirect to errors url because we set the website for the user
-      after_login_redirect
+      return after_login_redirect
     else
       # it will redirect to create website wizard because we don't sent it
-      after_login_redirect
+      return after_login_redirect
```
Comment (Chocksy) [inline app/controllers/invites_controller.rb]: why did you add return here? Was this necessary?

File: app/controllers/users_controller.rb [hunk 25]
```diff
@@ -2,6 +2,7 @@ class UsersController < ApplicationController
   layout 'landing', :only => [:new, :create]
   load_and_authorize_resource only: [:edit, :update]
   skip_before_action :authenticate!
+  attr_accessor :redirect_url
```
Comment (Chocksy) [inline app/controllers/users_controller.rb]: attr_accessor in a controller this is something i never saw... Why did you add this it's not even something you need...

File: app/controllers/users_controller.rb [hunk 26]
```diff
@@ -23,7 +24,10 @@ def new
   def create
     user = User.new(user_params)
 
-    authenticate!(:password) if user.save
+    if user.save!
```
Comment (Chocksy) [inline app/controllers/users_controller.rb]: don't let user.save! here as you'll get an exception for validation errors. You can put an else if you want to show an error when trying to sign up. That would be much better probably.

File: app/controllers/invites_controller.rb (hunk 24 above)
Comment (Panioglo) [inline app/controllers/invites_controller.rb]: it was my mistake, there is no need for a return

File: app/controllers/users_controller.rb [hunk 27]
```diff
@@ -13,16 +21,34 @@ def new
...
+    redirect_url = nil
+
+    if @user.save
+      authenticate!(:password)
+      @user.send_confirmation
+    else
+      redirect_to(signup_url, notice: "There was a problem") && return
+    end
+    redirect_url = accept_invite_url(params[:token]) if params[:token].present?
+    after_login_redirect(redirect_url)
```
Comment (Chocksy) [inline app/controllers/users_controller.rb]: move lines 33 and 34 in the if statement and then in the else write render :new to show the form again with the filled details and errors for each input.

File: app/views/users/new.html.haml [hunk 28]
```diff
@@ -12,6 +12,9 @@
               Sign up with Github
           %span.or or
           = form_for @user, url: signup_path do |f|
```
Comment (Chocksy) [inline app/views/users/new.html.haml]: if you use `simple_form_for @user` here you should have errors for your inputs displayed if there is a problem. Do that.

File: app/mailers/user_mailer.rb [hunk 29]
```diff
@@ -9,12 +15,11 @@ def notify_subscriber(group, user, message)
...
 
-  def member_invitation(website_id, email, website_member_id, inviter_id)
-    @user = User.find(inviter_id)
-    @website = Website.find(website_id)
-    @token = WebsiteMember.find(website_member_id).invitation_token
-    @email = email
-    mail(to: @email, subject: 'Epic Logger Invite Users')
+  def member_invitation(invite)
+    @invite = invite
+    @inviter = User.find(@invite.invited_by_id)
```
Comment (Chocksy) [inline app/mailers/user_mailer.rb]: This is very dull. :D So you need to add to invite.rb this:
`belongs_to :inviter, foreign_key: :invited_by_id, class_name: User` that way you can then do invite.inviter and get the right inviter for your invite.

File: app/models/invite.rb [hunk 30]
```diff
@@ -0,0 +1,20 @@
+class Invite < ActiveRecord::Base
+  # include the TokenGenerator extension
+  include TokenGenerator
+  belongs_to :website
```
Comment (Chocksy) [inline app/models/invite.rb:4]: so add this here:
`belongs_to :inviter, foreign_key: :invited_by_id, class_name: User`

File: app/mailers/user_mailer.rb (hunk 29 above)
Comment (Chocksy) [inline app/mailers/user_mailer.rb]: You also need to add a `has_many` clause in the user.rb model.
http://apidock.com/rails/ActiveRecord/Associations/ClassMethods/belongs_to
http://apidock.com/rails/v4.2.1/ActiveRecord/Associations/ClassMethods/has_many

Comment (Chocksy) [inline app/mailers/user_mailer.rb]: The has many can be like this:
`has_many :invites, foreign_key: :invited_by_id`

File: spec/controllers/invites_controller_spec.rb [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+  before(:each) do session[:epiclogger_website_id] = website.id end
+  # before(:each) do { expect( controller.class.skip_before_action ).to receive(:authenticate!).and_return(true) } end
+
+  describe 'POST #create' do
+    let(:params) { default_params.merge({invite: { email: 'invited@person.com' }}) }
+
+    it 'should raise error' do
+      create :invite, website: website, email: 'invited@person.com'
+      expect{ post_with user, params }.to raise_error
+      # expect{ post_with user, params }.to raise_exception(InvitesController, "Duplicate invites")
```
Comment (Chocksy) [inline spec/controllers/invites_controller_spec.rb]: let's remove comments form our code. They are not needed.

File: spec/controllers/website_members_controller_spec.rb [hunk 32]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe WebsiteMembersController, :type => :controller do
+	let(:user) { create :user }
```
Comment (Chocksy) [inline spec/controllers/website_members_controller_spec.rb]: weird indentation here.

File: app/controllers/users_controller.rb [hunk 33]
```diff
@@ -13,16 +21,34 @@ def new
...
+    @user = User.new(user_params)
+    redirect_url = nil
+
+    if @user.save
+      authenticate!(:password_unconfirmed)
+      @user.send_confirmation
+      redirect_url = accept_invite_url(params[:token]) if params[:token].present?
+      after_login_redirect(redirect_url)
+    else
+      render :new
```
Comment (Chocksy) [inline app/controllers/users_controller.rb:33]: You need to add specs so we check what happens when the user can't be saved. For example test when you pass a short password and check that we render new.
